### PR TITLE
feat(frontend): enable mm UUIDs for all modalities

### PR DESCRIPTION
**Pull request by Jankwi**
## Overview:
Previously, dynamo forwarded image UUIDs to vLLM, but explicitly blocked other modalities.
However, vLLM already supports using client-provided multi-modal UUIDs for video and audio.
This PR adds this capability to dynamo vLLM-based deployments.

## Details:
Previously, dynamo allowed multi-modal UUIDs only in image requests to vLLM.

This PR drops the modality check from that guard, and always forwards valid UUIDs to vLLM for all modalities.

UUID-based requests containing empty URLs for non-image modalities are still blocked, as that functionality is not yet supported for videos in vLLM 0.28.0.

## Where should the reviewer start?

***Frontend***: `components/src/dynamo/frontend/utils.py`, `components/src/dynamo/frontend/vllm_processor.py`, `lib/llm/src/http/service/openai.rs`, `lib/llm/src/preprocessor.rs`,

***Worker***: `components/src/dynamo/vllm/multimodal_utils/request_processor.py`,

***Sidecar***: `lib/sidecar/vllm/src/convert.rs`

### Frontend

```bash
python3 -m pytest -q \
  components/src/dynamo/common/tests/multimodal/test_cache_uuid.py \
  components/src/dynamo/frontend/tests/test_multimodal_utils.py \
  components/src/dynamo/frontend/tests/test_vllm_processor_unit.py \
  components/src/dynamo/vllm/tests/test_vllm_renderer_api.py
```

```bash
cargo test -p dynamo-llm --no-default-features \
  --test preprocessor cached_multimodal_uuid

cargo test -p dynamo-llm --no-default-features \
  --lib test_parse_chat_completion_request_accepts_media_url_with_uuid

cargo test -p dynamo-llm --no-default-features \
  --lib test_parse_chat_completion_request_accepts_empty_image_url_with_uuid
```

### Worker
```bash
python3 -m pytest -q \
  components/src/dynamo/vllm/tests/multimodal_utils/test_vllm_request_processor.py
```

### Sidecar
```bash
cargo test -p dynamo-vllm-sidecar --lib \
  tests::mixed_multimodal_media_forwards_uuids_for_each_modality -- --exact

cargo test -p dynamo-vllm-sidecar --lib \
  tests::image_hash_is_used_as_cache_uuid_when_user_uuid_is_missing -- --exact```
```

## Related Issues
**🚫 This PR is NOT linked to an issue**:
- [x] Confirmed — no related issue

## Summary by CodeRabbit

* **New Features**
  * Multimodal cache UUIDs are now preserved and forwarded for URL-backed image, video, and audio inputs.
  * Image cache entries can still be reused using UUID-only references.
  * Added support for tracking UUIDs across multimodal requests and aggregated processing.

* **Bug Fixes**
  * Audio- and video-only UUID references now return clear modality-specific validation errors and require a media URL.

* **Documentation**
  * Updated vLLM multimodal documentation to describe UUID forwarding and image-only UUID reuse.

### CODE REVIEW

File: components/src/dynamo/vllm/multimodal_utils/request_processor.py [hunk 1]
```diff
@@ -146,43 +146,36 @@ def _build_user_mm_uuids(
...
+            backend_modality = backend_modality.removesuffix("_url")
+        backend_modality = _normalize_forwarded_mm_modality(
+            backend_modality,
+            use_unified_vision_chunk,
         )
-        if has_uuid:
-            raise ValueError(
-                "multimodal cache UUIDs must use the 'image_url' modality key"
-            )
+        mm_uuids.setdefault(backend_modality, []).extend(values)
```
Comment (rmccorm4) [inline components/src/dynamo/vllm/multimodal_utils/request_processor.py:181]: I think we should omit modalities that do not contain a user-provided UUID. Otherwise, a request with an image UUID, standalone audio, and `use_audio_in_video=true` sends one audio UUID entry for two audio items, which vLLM rejects as a length mismatch. We still need to preserve `None` placeholders within modalities that contain at least one UUID.

File: components/src/dynamo/frontend/tests/test_multimodal_utils.py [hunk 2]
```diff
@@ -212,15 +212,44 @@ def test_extracts_url_and_uuid_only_slots_with_alignment():
     )
 
 
+@pytest.mark.parametrize("part_type", ["image_url", "audio_url", "video_url"])
```
Comment (rmccorm4) [inline components/src/dynamo/frontend/tests/test_multimodal_utils.py]: Could we remove `image_url` from this parametrization? The preceding test already covers image UUID extraction and alignment; the new behavior here is forwarding audio and video UUIDs.

File: components/src/dynamo/vllm/multimodal_utils/request_processor.py [hunk 3]
```diff
@@ -146,43 +146,36 @@ def _build_user_mm_uuids(
...
-                f"strings or null; got invalid entry at index {index}"
-            )
-    if not any(value is not None for value in image_uuids):
-        return None
-    backend_modality = _normalize_forwarded_mm_modality(
-        "image",
-        use_unified_vision_chunk,
-    )
-    return {backend_modality: list(image_uuids)}
+    # If at least one UUID is present, return a list for each modality.
```
Comment (rmccorm4) [inline components/src/dynamo/vllm/multimodal_utils/request_processor.py]: Can we drop this comment? The return immediately below already makes the behavior clear.

File: components/src/dynamo/vllm/multimodal_utils/request_processor.py (hunk 1 above)
Comment (Jankwi) [reply components/src/dynamo/vllm/multimodal_utils/request_processor.py:181]: Done, adapted `test_build_tokens_prompt_forwards_user_uuids_for_each_modality` to the new contract.

File: lib/sidecar/vllm/README.md [hunk 4]
```diff
@@ -29,11 +29,11 @@ It is a standalone Rust executable and is also compiled into
 - Opaque `kv_transfer_params` handoff
 - Data-parallel rank routing and KV-event source discovery
 - Capability-gated RL pause/resume, sleep/wake, weight-transfer, and weight-version controls through native gRPC
-- Image, video, and audio URL and data-URI inputs; cache UUIDs remain image-only
+- Image, video, and audio URL and data-URI inputs with cache UUID forwarding
```
[the remark this answers is not in the document]
Comment (Jankwi) [reply lib/sidecar/vllm/README.md]: Right, vLLM 0.28.0's default path supports passing UUIDs for audio and video but its gRPC service does not. Reverting all sidecar related changes from this PR.

**Comment by Jankwi** [comment]
I have reverted all sidecar changes from this PR.

Though the vLLM 0.28.0 default path supports passing UUIDs for non-image modalities, its gRPC service does not.
From what I see, a [PR](https://github.com/vllm-project/vllm/pull/53760) which enables audio/video UUID support there has been already merged into upstream vLLM.

cc: @connorcarpenter15

File: components/src/dynamo/vllm/multimodal_utils/request_processor.py [hunk 5]
```diff
@@ -146,43 +146,38 @@ def _build_user_mm_uuids(
...
-                f"strings or null; got invalid entry at index {index}"
-            )
-    if not any(value is not None for value in image_uuids):
-        return None
-    backend_modality = _normalize_forwarded_mm_modality(
-        "image",
-        use_unified_vision_chunk,
-    )
-    return {backend_modality: list(image_uuids)}
+        backend_modality = str(modality)
```
[the remark this answers is not in the document]
Comment (Jankwi) [reply components/src/dynamo/vllm/multimodal_utils/request_processor.py:174]: Added coverage in `components/src/dynamo/vllm/tests/multimodal_utils/test_vllm_request_processor.py::test_audio_in_video_aligns_derived_audio_uuids` and fixed

**Comment by Jankwi** [comment]
When `use_audio_in_video` is set, audio derived from video inherits its UUID (in line with vLLM).
There are many edge cases, happy to hear your feedback.

**Review by rmccorm4 — APPROVED** [review]

**Comment by rmccorm4** [comment]
/ok to test [9c33b2f](https://github.com/ai-dynamo/dynamo/pull/14300/commits/9c33b2f23b7170d0d709e1a7a408654a4bc412b1)

File: components/src/dynamo/frontend/utils.py [hunk 6]
```diff
@@ -184,25 +183,23 @@ def extract_mm_urls(
...
-                    "image_url parts with vLLM"
-                )
-
             url = media_value.get("url") if isinstance(media_value, dict) else None
             if isinstance(url, str) and url:
                 mm_data.setdefault(part_type, []).append({"Url": url})
-            elif isinstance(uuid_value, str):
+            elif isinstance(uuid_value, str) and part_type == "image_url":
                 mm_data.setdefault(part_type, []).append({"UuidOnly": uuid_value})
+            elif isinstance(uuid_value, str):
```
Comment (GuanLuo) [inline components/src/dynamo/frontend/utils.py:189]: Can you combine with above to be, so that we are clear that the two cases are checked when uuid is provided
```
elif isinstance(uuid_value, str):
    if part_type == "image_url":
        mm_data.setdefault(part_type, []).append({"UuidOnly": uuid_value})
    else:
        raise ...
```

**Comment by rmccorm4** [comment]
/ok to test [5fa5300](https://github.com/ai-dynamo/dynamo/pull/14300/commits/5fa53009423a9b90e0b7abd06c1158a51731116f)

**Review by GuanLuo — APPROVED** [review]
LGTM with nit comment
